src/inventorymanager.cpp

```diff
@@ -471,69 +474,72 @@ void IMoveAction::apply(InventoryManager *mgr, ServerActiveObject *player, IGame
 	*/
 
-	/* Detached inventories */
-
-	// Both endpoints are same detached
-	if (from_inv.type == InventoryLocation::DETACHED &&
-			from_inv == to_inv) {
-		PLAYER_TO_SA(player)->detached_inventory_OnMove(
+	// Source = destination => move
+	if (from_inv == to_inv) {
+		if (from_inv.type == InventoryLocation::DETACHED)
+			PLAYER_TO_SA(player)->detached_inventory_OnMove(
 				*this, count, player);
+		else if (from_inv.type == InventoryLocation::NODEMETA)
+			PLAYER_TO_SA(player)->nodemeta_inventory_OnMove(
+				*this, count, player);
+		else if (from_inv.type == InventoryLocation::PLAYER)
+			PLAYER_TO_SA(player)->player_inventory_OnMove(
+				*this, count, player);
+		mgr->setInventoryModified(from_inv);
 	} else {
-		// Destination is detached
-		if (to_inv.type == InventoryLocation::DETACHED) {
+		// Destinations
+		if (to_inv.type == InventoryLocation::DETACHED)
 			PLAYER_TO_SA(player)->detached_inventory_OnPut(
 				*this, src_item, player);
-		}
-		// Source is detached
-		if (from_inv.type == InventoryLocation::DETACHED) {
-			PLAYER_TO_SA(player)->detached_inventory_OnTake(
-				*this, src_item, player);
-		}
-	}
-
-	/* Node metadata inventories */
-
-	// Both endpoints are same nodemeta
-	if (from_inv.type == InventoryLocation::NODEMETA &&
-			from_inv == to_inv) {
-		PLAYER_TO_SA(player)->nodemeta_inventory_OnMove(
-			*this, count, player);
-	} else {
-		// Destination is nodemeta
-		if (to_inv.type == InventoryLocation::NODEMETA) {
+		else if (to_inv.type == InventoryLocation::NODEMETA)
 			PLAYER_TO_SA(player)->nodemeta_inventory_OnPut(
 				*this, src_item, player);
-		}
-		// Source is nodemeta
-		if (from_inv.type == InventoryLocation::NODEMETA) {
-			PLAYER_TO_SA(player)->nodemeta_inventory_OnTake(
-				*this, src_item, player);
-		}
-	}
-
-	// Player inventories
-
-	// Both endpoints are same player inventory
-	if (from_inv.type == InventoryLocation::PLAYER &&
-			from_inv == to_inv) {
-		PLAYER_TO_SA(player)->player_inventory_OnMove(
-			*this, count, player);
-	} else {
-		// Destination is player inventory
-		if (to_inv.type == InventoryLocation::PLAYER) {
+		else if (to_inv.type == InventoryLocation::PLAYER)
 			PLAYER_TO_SA(player)->player_inventory_OnPut(
 				*this, src_item, player);
-		}
-		// Source is player inventory
-		if (from_inv.type == InventoryLocation::PLAYER) {
+		// Sources
+		if (from_inv.type == InventoryLocation::DETACHED)
+			PLAYER_TO_SA(player)->detached_inventory_OnTake(
+				*this, src_item, player);
+		else if (from_inv.type == InventoryLocation::NODEMETA)
+			PLAYER_TO_SA(player)->nodemeta_inventory_OnTake(
+				*this, src_item, player);
+		else if (from_inv.type == InventoryLocation::PLAYER)
 			PLAYER_TO_SA(player)->player_inventory_OnTake(
 				*this, src_item, player);
+		if (did_swap) {
+			// Item is now placed in source list
+			src_item = list_from->getItem(from_i);
+			std::swap(from_inv, to_inv);
+			std::swap(from_list, to_list);
+			std::swap(from_i, to_i);
```

Comment:
Please move this to a new function, such as `IMoveAction::swapDirection` or similar.
The advantage of such a function would be to reduce confusions and make the code less error-prone (especially when swapping back and forth).